File: submissions/0270-closest-binary-search-tree-value/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function closestValue(root: TreeNode | null, target: number): number {
        const inorder = (r: TreeNode | null): number[] => {
            return r ? [...inorder(r.left), r.val, ...inorder(r.right)] : [];
        };

        const values = inorder(root);
        return values.reduce((closest, x) => Math.abs(target - x) < Math.abs(target - closest) ? x : closest, values[0]);
    }

File: submissions/0270-closest-binary-search-tree-value/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def closestValue(self, root: Optional[TreeNode], target: float) -> int:    
        def dfs(node, ans):
            if not node:
                return ans
            
            if (abs(target - node.val) < abs(target - ans)) or (
                abs(target - node.val) == abs(target - ans) and node.val < ans):
                ans = node.val
        
            if node.val > target:
                return dfs(node.left, ans)
            elif node.val < target:
                return dfs(node.right, ans)
            else:
                return node.val

        
        return dfs(root, root.val)
        
